src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs

```diff
@@ -0,0 +1,182 @@
+                    context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                    instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                    instance.ClientFactory.GetCustomHandlers());
+
+                client.SubscriptionId = context.Subscription.Id;
+
+                var odata = new ODataQuery<GenericResourceFilter>(r => r.ResourceType == resourceType);
+
+                IEnumerable<ResourceInfo> resourceInfoList;
+                try
+                {
+                    resourceInfoList = client.Resources.List(odata)
+                        .Select(r => ResourceInfo.FromResourceId(r.Id))
+                        .ToList();
+                }
+                catch (ArgumentException)
+                {
+                    return new List<ResourceInfo>();
+                }
+
+                if (cacheItem != null)
+                {
+                    cacheItem.Timestamp = DateTime.Now;
+                    cacheItem.ResourceInfoList = resourceInfoList;
+                    Cache[contextHash] = cacheItem;
+                }
+                else
+                {
+                    Cache.Add(contextHash, new CacheItem
+                    {
+                        Timestamp = DateTime.Now,
+                        ResourceInfoList = resourceInfoList
+                    });
+                }
+
+                return resourceInfoList;
+            }
+        }
+
+        private static int HashContext(IAzureContext context, string resourceType)
+        {
+            return (context.Account.Id + context.Environment.Name + context.Subscription.Id + context.Tenant.Id + resourceType).GetHashCode();
+        }
+
+        public partial class ResourceInfo
+        {
+            private static class Patterns
+            {
+                // example: /subscriptions/1c638cf4-608f-4ee6-b680-c329e824c3a8/resourceGroups/rgrssdfw69c0219330/providers/Microsoft.Sql/servers/sqlserver2c5254480/databases/master
+                private const string SubscripitonId = @"([0-9a-f]{8}[-][0-9a-f]{4}[-][0-9a-f]{4}[-][0-9a-f]{4}[-][0-9a-f]{12})";
+                private const string ResourceGroupName = @"([^/]+)";
+                private const string ProviderName = @"(Microsoft\.[^/]+)";
+                private const string ServiceName = @"([^/]+)";
+                private const string ResourceName = @"([^/]+)";
+                private const string Sub = @"(/[^/]+/[^/]+)?";
+
+                public static readonly string ResourceId = $"/subscriptions/{SubscripitonId}/resourceGroups/{ResourceGroupName}/providers/{ProviderName}/{ServiceName}/{ResourceName}{Sub}";
+            }
+
+            public static ResourceInfo FromResourceId(string resourceId)
+            {
+                var match = Regex.Match(resourceId, Patterns.ResourceId, RegexOptions.IgnoreCase);
```

Comment:
The numbers here seem a little magically to me, but this is fine as long as you've tested it on resources at multiple levels.

Reply:
2 levels only